File: backend-app/src/main/java/com/backend_app/backend_app/service/TransaccionService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Transaccion;
import com.backend_app.backend_app.dto.TransaccionDTO;
import com.backend_app.backend_app.dto.UsuarioDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface TransaccionService {

    List<TransaccionDTO> listadoTransacciones();

    Optional<TransaccionDTO> buscarTransaccionPorId(long idTransaccioin);

    TransaccionDTO convertirADTO(Transaccion transaccion);

    void registrarTransaccion(TransaccionDTO transaccionDTO);

    void actualizarTransaccion(TransaccionDTO transaccionDTO);

    void eliminarTransaccion(TransaccionDTO transaccionDTO);

    Page<TransaccionDTO> findAll(Pageable pageable);

    Integer cantidadTransacciones(Long usuarioId);

    List<TransaccionDTO> top5Transacciones();

    List<TransaccionDTO> historialTransacciones(Long idUsuario);

    List<TransaccionDTO> historialTransaccionesVM();

    List<TransaccionDTO> listadoTransaccionesCompletos();

}

File: frontend-app/src/app/services/transaccion.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Transaccion } from "../domain/Transaccion";
import { Usuario } from "../domain/Usuario";

@Injectable({
    providedIn: 'root'
})
export class TransaccionService{

    private url: string = 'http://localhost:8080/api/transacciones';

    constructor(private http: HttpClient){

    }

    listadoTransacciones(): Observable<Transaccion[]>{
        return this.http.get<Transaccion[]>(this.url);
    }

    buscarTransaccionPorId(id: number): Observable<Transaccion>{
        return this.http.get<Transaccion>(`${this.url}/${id}`);
    }

    registrarUsuario(usuario: any): Observable<any>{
        return this.http.post('http://localhost:8080/api/usuarios/guardar', usuario);
    }


    guardarTransaccion(transaccion: Transaccion){
        return this.http.post('http://localhost:8080/api/transacciones/guardar', transaccion);
    }
    
    editarTransaccion(transaccion: Transaccion) {
        return this.http.put(`${this.url}/actualizar`, transaccion);
    }
    
    eliminarTransaccion(idTransaccion: number){
        return this.http.delete(`${this.url}/${idTransaccion}`);
    }

    listadoTransaccionesPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }
    
    getCantidadTransacciones(usuarioId: number): Observable<number> {
        return this.http.get<number>(
            `${this.url}/cantidad/${usuarioId}`
        );
    }

    top5Transacciones(): Observable<Transaccion[]> {
        return this.http.get<Transaccion[]>(`${this.url}/top5`);
    }

    historialTransacciones(idUsuario: number): Observable<Transaccion[]> {
        return this.http.get<Transaccion[]>(`${this.url}/historial/${idUsuario}`);
    }

    historialTransaccionesVM(): Observable<Transaccion[]>{
        return this.http.get<Transaccion[]>(`${this.url}/historial-transacciones`);
    }

    listadoTransaccionesCompletos(): Observable<Transaccion[]>{
        return this.http.get<Transaccion[]>(`${this.url}/completos`);
    }

}